tool/tsh/aliases.go

```diff
@@ -0,0 +1,176 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/shlex"
+	"github.com/gravitational/trace"
+)
+
+const tshAliasEnvKey = "TSH_ALIAS"
+
+// aliasRunner coordinates alias running as well as provides a suitable testing target.
+type aliasRunner struct {
+	getEnv func(key string) string
+	setEnv func(key, value string) error
+
+	aliases map[string]string
+
+	runTshMain         func(ctx context.Context, args []string, opts ...cliOption) error
+	runExternalCommand func(cmd *exec.Cmd) error
+}
+
+// newAliasRunner returns regular alias runner; the tests are using a different function.
+func newAliasRunner(aliases map[string]string) *aliasRunner {
+	return &aliasRunner{
+		getEnv:     os.Getenv,
+		setEnv:     os.Setenv,
+		aliases:    aliases,
+		runTshMain: Run,
+		runExternalCommand: func(cmd *exec.Cmd) error {
+			return cmd.Run()
+		},
+	}
+}
+
+// findAliasCommand inspects the argument list to find first non-option (i.e. command). The command is returned along with the argument list from which the command was removed.
+func findAliasCommand(args []string) (string, []string) {
+	aliasCmd := ""
+	aliasIx := -1
+
+	for i, arg := range args {
+		if arg == "" {
+			continue
+		}
+
+		if strings.HasPrefix(arg, "-") {
+			continue
+		}
+
+		aliasCmd = arg
+		aliasIx = i
+		break
+	}
+
+	if aliasCmd == "" {
+		return "", nil
+	}
+
+	runtimeArgs := make([]string, 0)
+	runtimeArgs = append(runtimeArgs, args[:aliasIx]...)
+	runtimeArgs = append(runtimeArgs, args[aliasIx+1:]...)
+
+	return aliasCmd, runtimeArgs
+}
+
+// expandAliasDefinition expands $0, $1, ... within alias definition. Arguments not referenced in alias are appended at the end.
+func expandAliasDefinition(aliasDef string, runtimeArgs []string) ([]string, error) {
+	var appendArgs []string
+
+	for i, arg := range runtimeArgs {
+		variable := "$" + strconv.Itoa(i)
+		if strings.Contains(aliasDef, variable) {
+			aliasDef = strings.ReplaceAll(aliasDef, variable, arg)
+		} else {
+			appendArgs = append(appendArgs, arg)
+		}
+	}
+
+	missingRefs := regexp.MustCompile(`\$\d`)
+	if missingRefs.MatchString(aliasDef) {
+		return nil, trace.BadParameter("unsatisfied argument variables for alias: %v", aliasDef)
+	}
+
+	split, err := shlex.Split(aliasDef)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	out := append(split, appendArgs...)
+	return out, nil
+}
+
+// getAliasDefinition returns the alias definition if it exists and the alias is still eligible for running.
+func (ar *aliasRunner) getAliasDefinition(aliasCmd string) (bool, string) {
```

Comment:
nit: Can we reverse the returned args to "string"," bool" to be more complied with common GO convention ?

Reply:
Done.